### ISSUE TITLE
Page breaks are unclean when outputting paginated PDFs of transcripts in Marked

### ISSUE THREAD

**Issue by briandk** [briandk/transdown#32]
## Bug report overview

- What the problem looks like
- What I expect should happen
- What actually happens
- How to Reproduce the Bug
	- Notes on reproducing
	- Steps to reproduce
- Notes to the developer (Brett Terpstra)

## What the Problem Looks Like

Marked2 (hereafter: Marked) renders Transdown files beautifully. But, when I use Marked’s export function to output a paginated PDF:

![How I output Paginated PDFs in Marked](http://f.cl.ly/items/151D120b251U362a3s1L/RebeccaWellsFrameShiftEpisode20120406_md.png)

The page breaks aren’t always clean:

![Unclean Page Break example 1, where a line is split lengthwise across two pages](http://f.cl.ly/items/1M3O002D160K3C421L06/RebeccaWellsFrameShiftEpisode20120406-1_pdf__page_2_of_5_.png)

![Example of an unclean page break on a longer line](http://f.cl.ly/items/3r3O1w3V1M2b0O3g3M3M/RebeccaWellsFrameShiftEpisode20120406-1_pdf__page_3_of_5_.png)

## What I Expect Should Happen

I guess what I *hope* would happen is that lines would be broken cleanly and appropriately across the page. 

## What Actually Happens

Unclean page breaks like these:

![Unclean Page Break example 1, where a line is split lengthwise across two pages](http://f.cl.ly/items/1M3O002D160K3C421L06/RebeccaWellsFrameShiftEpisode20120406-1_pdf__page_2_of_5_.png)

![Example of an unclean page break on a longer line](http://f.cl.ly/items/3r3O1w3V1M2b0O3g3M3M/RebeccaWellsFrameShiftEpisode20120406-1_pdf__page_3_of_5_.png)

### Notes on reproducing

1. All steps involving the `transdown` directory may require you to specify your full (absolute) path to the file in question on your machine. So, I’ve just specified the path relative to the root of your repository.
2. My current export settings are: ![Screenshot of my current export settings](http://f.cl.ly/items/3t3d0C0x2u2C0h1e082h/Export.png)

### Steps to reproduce

1. Clone a copy of Transdown: <https://github.com/briandk/transdown>
2. Checkout the `marked-pagination-bug` branch in Git
3. In Marked, set your Custom Processor settings as follows
	1. **Path**: Your Node path, likely `usr/local/bin/node`
	2. **Args**: The full path to this script: `transdown/javascripts/renderMarkdownPreviewInMarked.js`
4. In Marked, add the following custom CSS file: `transdown/stylesheets/markedApp/transdownMarkedPreview.css`
5. In Marked, load this Markdown example file from the transdown repo: `transdown/markedAppDemo/RebeccaWellsFrameShiftEpisode20120406.md`
6. In the Marked preview window, toggle on your custom processor. 
7. In the Marked preview window, switch your CSS style to `transdownMarkedPreview.css`
8. In the Marked preview window, hit the Export Drawer icon, choose PDF (Paginated), and save your PDF file
9. Open the saved PDF and verify the breaks are unclean between pages 1–2 and pages 3–5

### Notes to the Developer (Brett Terpstra)

Honestly, I’m amazed this is even working. Having this functionality was a dream of mine. More pragmatically, though:

1. What I’m asking Marked to render is an HTML table, so that may be why it’s harder to judge page breaks.
2. The HTML I coded in [my Handlebars template][1] may be broken. (I am not an HTML wizard; I’m just doing my best).
3. I’m kind of star-struck that you’re looking into this. Thanks! 

[1]: https://github.com/briandk/transdown/blob/ec7683e3c5a5c2d25413eb6fe8a1d47b98519f98/transcriptTemplate.handlebars

**Comment by briandk** [comment]
Duh, forgot key system information:

- Mac OS X 10.10.2
- Marked2 Version 2.4.9 (876)

**Comment by ttscoff** [comment]
This is a great report. Tables can definitely be problematic to print, but I'd like to look for a solution.

**Comment by briandk** [comment]
@ttscoff - Quick update: I [merged the `marked-pagination-bug`][1] branch into master, but I also gave that commit a tag of `marked-pagination-bug`. 

So, this should still work as before:

```bash
> git checkout marked-pagination-bug
```

And the steps to reproduce are unchanged.

[1]: https://github.com/briandk/transdown/commit/3224be9156326a16f87d7fafe378a5e3560477c9

**Comment by ttscoff** [comment]
perfect, thanks!

**Comment by ttscoff** [comment]
WebKit does not make it easy to prevent breaking a `<tr>`. I'm wondering if you might try creating a new table for each row instead of all rows in one table... horrible HTML, but it would allow you to keep the same styling but use a `table { page-break-inside:avoid; }` rule to jump the entire row to the next page when it would be cut off...

**Comment by briandk** [comment]
@ttscoff - Again, thanks for looking into this!

1. I can give the one-table-per-row thing a shot, but it would change my current spec (which allocates [one table per episode of transcript][2]).
2. I did not know CSS had a [`page-break-inside`][1] property! Cool!
3. On a whim, I tried this CSS `tr { page-break-inside:avoid; }`. But, it has no effect, presumably because that CSS property only applies to `<p>` elements?
4. I think I see what you mean about WebKit not making it easy. I tried [some][3] [StackOverflow][4] [threads][5], but nothing quite worked. I'm guessing this is an area where not all browsers implement things the same way? Also, I'm guessing Marked uses something like a [`WKWebView`][6] to render preview content, which means the app is bound to whatever restrictions WebKit has implemented for handling this particular property and circumstance.[^a]


[^a]: Disclaimer - I am also neither an [NSHipster][7] nor a Swift Ninja, so I'm really guessing here.

[1]: http://www.w3schools.com/cssref/pr_print_pagebi.asp
[2]: https://github.com/briandk/transdown/blob/master/transcriptTemplate.handlebars#L1-L4
[3]: http://stackoverflow.com/questions/8712677/how-to-apply-css-page-break-to-print-a-table-with-lots-of-rows
[4]: http://stackoverflow.com/questions/9288802/avoid-page-break-inside-row-of-table
[5]: http://stackoverflow.com/questions/1763639/how-to-deal-with-page-breaks-when-printing-a-large-html-table
[6]: https://developer.apple.com/library/ios/documentation/WebKit/Reference/WKWebView_Ref/index.html#//apple_ref/occ/cl/WKWebView
[7]: http://nshipster.com/wkwebkit/

**Comment by ttscoff** [comment]
1. I didn't actually get a chance yet to try it, but I'm 99% certain it's the easiest path without coming up with a whole new stylesheet to work with list-style markup
2. page-break-[inside, before, after] is handy but a bit unpredictable between browsers
3. Yeah, that was the first thing I tried. I also tried wrapping the `<td>` inner content in `<div>` and applying the page-break-inside to those, but no luck.
4. This is definitely a cross-browser annoyance. Firefox handles number 3 well. WebKit also ignores any @page directives, and breaks intra-document links when printing PDFs.

Marked uses WebView, but I haven't updated it to the newer WkWebView yet because it didn't fix any of the major problems in my testing. It's still straight up WebKit, though, and equivalent to Safari's version. Chrome's branch (Blink) has some benefits, and there are QT versions of WebKit, but I'm not embedding any third party, 100+MB APIs in the app :).

**Comment by briandk** [comment]
## tl;dr

To me, this page break issue clearly isn't a Marked bug, and it might be an edge case my users are willing to live with.

## Slightly longer version

@ttscoff - So, this problem has two features worth thinking about:

1. In their current state, transdown documents don't page-break nicely when outputted as PDFs. BUT, that may not actually be the most important use-case for my users.
2. The page-break problem is a cross-browser issue, and in NO way am I suggesting you change anything about Marked to fix it :-). Heck, the problem might go away on its own with future updates to `WebView` or `WkWebView`. Or at least, that's a possibility.

So, in sum I'm thinking ***this problem might be an edge case***. I think for now my users would be willing to put up with occasional weird linebreaks in a PDF, because most of them won't necessarily be outputting paginated PDFs anyway. More likely they'll be outputting monolithic HTML files with everything embedded:

![save_as_html](https://cloud.githubusercontent.com/assets/330036/6378392/408796c0-bce8-11e4-997a-5937e01ccb0b.png)

And, I think it's pretty clearly not at all a bug with Marked. It's a consequence of my (for now) chosen HTML structure of `<tr>`s and WebKit's particular handling of their styles.
